programs/mango-v4/tests/test_reduce_only.rs

```diff
@@ -0,0 +1,571 @@
+    let bank = solana.get_account::<Bank>(tokens[0].bank).await;
+    let native = mango_account_0.tokens[0].native(&bank);
+    assert_eq!(native.to_num::<u64>(), initial_token_deposit);
+
+    // withdraw all should pass
+    send_tx(
+        solana,
+        TokenWithdrawInstruction {
+            amount: initial_token_deposit,
+            allow_borrow: false,
+            account: account_0,
+            owner,
+            token_account: payer_mint_accounts[0],
+            bank_index: 0,
+        },
+    )
+    .await
+    .unwrap();
+
+    // borrowing should fail
+    let res = send_tx(
+        solana,
+        TokenWithdrawInstruction {
+            amount: initial_token_deposit,
+            allow_borrow: true,
+            account: account_0,
+            owner,
+            token_account: payer_mint_accounts[0],
+            bank_index: 0,
+        },
+    )
+    .await;
+    assert!(res.is_err());
+
+    Ok(())
+}
+
+#[tokio::test]
+async fn test_perp_reduce_only() -> Result<(), TransportError> {
+    let context = TestContext::new().await;
+    let solana = &context.solana.clone();
+
+    let admin = TestKeypair::new();
+    let owner = context.users[0].key;
+    let payer = context.users[1].key;
+    let mints = &context.mints[0..=2];
+
+    let initial_token_deposit = 1000_000;
+
+    //
+    // SETUP: Create a group and an account
+    //
+
+    let GroupWithTokens { group, tokens, .. } = GroupWithTokensConfig {
+        admin,
+        payer,
+        mints: mints.to_vec(),
+        ..GroupWithTokensConfig::default()
+    }
+    .create(solana)
+    .await;
+
+    let settler =
+        create_funded_account(&solana, group, owner, 251, &context.users[1], &[], 0, 0).await;
+    let settler_owner = owner.clone();
+
+    let account_0 = create_funded_account(
+        &solana,
+        group,
+        owner,
+        0,
+        &context.users[1],
+        &mints[0..1],
+        initial_token_deposit,
+        0,
+    )
+    .await;
+    let account_1 = create_funded_account(
+        &solana,
+        group,
+        owner,
+        1,
+        &context.users[1],
+        &mints[0..1],
+        initial_token_deposit * 100, // Fund 100x, so that this is not the bound for what account_0 can settle
+        0,
+    )
+    .await;
+
+    //
+    // TEST: Create a perp market
+    //
+    let mango_v4::accounts::PerpCreateMarket { perp_market, .. } = send_tx(
+        solana,
+        PerpCreateMarketInstruction {
+            group,
+            admin,
+            payer,
+            perp_market_index: 0,
+            quote_lot_size: 10,
+            base_lot_size: 100,
+            maint_asset_weight: 0.975,
+            init_asset_weight: 0.95,
+            maint_liab_weight: 1.025,
+            init_liab_weight: 1.05,
+            liquidation_fee: 0.012,
+            maker_fee: 0.0002,
+            taker_fee: 0.000,
+            settle_pnl_limit_factor: 0.2,
```

Comment:
highly recommend you set this to -1 in tests that don't explicitly test settle limits, otherwise they become pretty complex (this test is fine, doesn't settle anything after all)